src/graph/network/communicator.h

```diff
@@ -6,112 +6,144 @@
+   * \param msg data message
    * \param recv_id receiver's ID
-   * \return bytes we sent
+   * \return bytes of data
    *   > 0 : bytes we sent
    *   - 1 : error
    */
-  virtual int64_t Send(const char* data, int64_t size, int recv_id) = 0;
+  virtual int64_t Send(Message msg, int recv_id) = 0;
 
   /*!
    * \brief Finalize Sender
    */
   virtual void Finalize() = 0;
 
   /*!
-   * \brief Get data buffer
-   * \return buffer pointer
+   * \brief Communicator type: 'socket', 'mpi', etc.
    */
-  virtual char* GetBuffer() = 0;
+  virtual std::string Type() const = 0;
 
+ protected:
   /*!
-   * \brief Set data buffer
+   * \brief Size of message queue
    */
-  virtual void SetBuffer(char* buffer) = 0;
+  int64_t queue_size_;
 };
 
 /*!
  * \brief Network Receiver for DGL distributed training.
  *
- * Receiver is an abstract class that defines a set of APIs for receiving binary 
- * data over network. It can be implemented by different underlying networking libraries 
- * such TCP socket and ZMQ. One Receiver can connect with multiple Senders, and it can receive 
- * data from these Senders concurrently via multi-threading and message queue.
+ * Receiver is an abstract class that defines a set of APIs for receiving binary data 
+ * message over network. It can be implemented by different underlying networking 
+ * libraries such as TCP socket and MPI. One Receiver can connect with multiple Senders 
+ * and it can receive data from multiple Senders concurrently.
  */
 class Receiver {
  public:
+  /*!
+   * \brief Receiver constructor
+   * \param queue_size size of message queue.
+   * Note that, the queue_size parameter is optional.
+   */
+  explicit Receiver(int64_t queue_size = 0) {
+    if (queue_size < 0) {
+      LOG(FATAL) << "queue_size cannot be a negative number.";
+    }
+    queue_size_ = queue_size;
+  }
+
   virtual ~Receiver() {}
 
   /*!
-   * \brief Wait all of the Senders to connect
-   * \param ip Receiver's IP address
-   * \param port Receiver's port
+   * \brief Wait for all the Senders to connect
+   * \param addr Networking address, e.g., 'socket://127.0.0.1:50051', 'mpi://0'
    * \param num_sender total number of Senders
-   * \param queue_size size of message queue
-   * \return True for sucess and False for fail
+   * \return True for success and False for fail
    */
-  virtual bool Wait(const char* ip, int port, int num_sender, int queue_size) = 0;
+  virtual bool Wait(const char* addr, int num_sender) = 0;
 
   /*!
-   * \brief Recv data from Sender (copy data from message queue)
-   * \param dest data buffer of destination
-   * \param max_size maximul size of data buffer
-   * \return bytes we received
-   *   > 0 : bytes we received
+   * \brief Recv data from Sender
+   * \param msg pointer of data message
+   * \param send_id which sender current msg comes from
+   * \return bytes of data
+   *   > 0 : bytes we sent
+   *   - 1 : error
+   */
+  virtual int64_t Recv(Message* msg, int* send_id) = 0;
```

Comment:
Similar to `Send`, update the docstring.
(1) Multi-thread safe
(2) Blocking
(3) Memory allocated by communicator but will not own it after the function returns.

Reply:
same